lib/deprecation.nix

```diff
@@ -47,14 +47,43 @@ rec {
         '';
     };
 
+  /*
+    Returns a function that maps
+      [
+        "someOption"
+        ["fooBar" "someSubOption"]
+        { old = "someOtherOption"; new = ["foo_bar" "some_other_option"]}
+      ]
+
+    to
+      [
+        (lib.mkRenamedOptionModule
+          (oldPath ++ ["someOption"])
+          (newPath ++ ["some_option"])
+        )
+        (lib.mkRenamedOptionModule
+          (oldPath ++ ["fooBar" "someSubOption"])
+          (newPath ++ ["foo_bar" "some_sub_option"])
+        )
+        (lib.mkRenamedOptionModule
+          (oldPath ++ ["someOtherOption"])
+          (newPath ++ ["foo_bar" "some_other_option"])
+        )
+      ]
+  */
   mkSettingsRenamedOptionModules =
     oldPrefix: newPrefix:
     map (
       option':
+      # If the value is an attrs, it must have both `old` and `new` attributes.
+      assert lib.isAttrs option' -> ((option' ? old) && (option' ? new));
       let
-        option = lib.toList option';
-        oldPath = oldPrefix ++ option;
-        newPath = newPrefix ++ map lib.nixvim.toSnakeCase option;
+        oldSubPath = lib.toList (option'.old or option');
+        newSubPath =
+          if lib.isAttrs option' then lib.toList option'.new else map lib.nixvim.toSnakeCase oldSubPath;
```

Comment:
How about flipping the logic a little by coercing non-attrs to a normalised form?

```nix
option'' =
  if isAttrs option' then
    mapAttrs (_: toList) option'
  else
    {
      old = toList option';
      new = map lib.nixvim.toSnakeCase option''.old;
    };
```

Maybe better names than `option'` and `option''` :grin:

Reply:
Maybe rename `option'` to `rename`, `spec`, or something?
And `option''` to `normalized`, `final` or something